test/gtest/ucp/test_ucp_request.cc

```diff
@@ -100,3 +103,489 @@ UCS_TEST_P(test_ucp_request, test_request_query)
+                               (uint64_t)m_sbuf.data(), m_rkey, &param);
+            break;
+
+        case STREAM:
+            param.op_attr_mask = UCP_OP_ATTR_FIELD_DATATYPE;
+            param.datatype     = ucp_dt_make_contig(1);
+            sreq = ucp_stream_send_nbx(sender().ep(), m_sbuf.data(), m_msg_size,
+                                       &param);
+            break;
+
+        case AM:
+            sreq = send_am();
+            break;
+        }
+
+        ASSERT_TRUE(UCS_PTR_IS_PTR(sreq));
+        wait_and_restart(sreq, rreq, output_proto);
+        wait_receive(op);
+
+        request_wait(sreq);
+        request_wait(rreq);
+        flush_ep(sender());
+
+        EXPECT_EQ(m_sbuf, m_rbuf);
+    }
+
+    static void restart_completed(ucp_ep_h ep, void *arg)
+    {
+        unsigned *completed_p = (unsigned*)arg;
+        *completed_p          = 1;
+    }
+
+    virtual void
+    wait_and_restart(void *sreq, void *rreq, const std::string &expected_proto)
+    {
+        ucp_request_t *req                 = (ucp_request_t*)sreq - 1;
+        const ucp_datatype_iter_t *dt_iter = &req->send.state.dt_iter;
+
+        while (dt_iter->offset == 0) {
+            progress();
+        }
+
+        EXPECT_LT(dt_iter->offset, dt_iter->length);
+        ASSERT_UCS_OK(ucp_ep_pending_schedule_restart(sender().ep(),
+                                                      restart_completed,
+                                                      &m_completed));
+
+        wait_for_value(&m_completed, (unsigned)1);
+        EXPECT_STREQ(expected_proto.c_str(),
+                     req->send.proto_config->proto->name);
+    }
+
+    void init_rkey()
+    {
+        const ucp_mem_map_params_t params = {
+            .field_mask = UCP_MEM_MAP_PARAM_FIELD_ADDRESS |
+                          UCP_MEM_MAP_PARAM_FIELD_LENGTH,
+            .address    = m_sbuf.data(),
+            .length     = m_sbuf.size()
+        };
+
+        ASSERT_UCS_OK(ucp_mem_map(receiver().ucph(), &params, &m_memh));
+
+        void *rkey_buffer;
+        size_t rkey_buffer_size;
+        ASSERT_UCS_OK(ucp_rkey_pack(receiver().ucph(), m_memh, &rkey_buffer,
+                                    &rkey_buffer_size));
+
+        ASSERT_UCS_OK(ucp_ep_rkey_unpack(sender().ep(), rkey_buffer, &m_rkey));
+        ucp_rkey_buffer_release(rkey_buffer);
+    }
+
+    static void get_test_variants(std::vector<ucp_test_variant> &variants)
+    {
+        add_variant_with_value(variants,
+                               UCP_FEATURE_TAG | UCP_FEATURE_RMA |
+                               UCP_FEATURE_STREAM | UCP_FEATURE_AM,
+                               0, "");
+    }
+
+    size_t m_msg_size;
+    std::vector<char> m_sbuf;
+    std::vector<char> m_rbuf;
+    unsigned m_completed;
+    ucp_mem_h m_memh;
+    ucp_rkey_h m_rkey;
+};
+
+UCS_TEST_P(test_proto_reset, eager_zcopy, "ZCOPY_THRESH=0", "RNDV_THRESH=inf")
+{
+    reset_protocol(TAG, "egr/multi/bcopy");
+}
```

Comment:
IMO better to keep test name the same as protocol name, in some cases current naming looks confusing

Reply:
test name is source protocol and argument name is dest protocol

Reply:
how does `eager_zcopy` match `egr/multi/bcopy`?

Reply:
as a first step we decided to support only bcopy protocols after reset

Reply:
Let's make naming of tests consistent to what it's testing

Reply:
you mean combine src +dst proto?

Reply:
what is src and dst? "sender and receiver" or switch "from to"? it's not clear without digging into implementation

Reply:
how about `eager_zcopy_to_eager_bcopy`?

Reply:
how to differentiate single/multi? IMO better to have full exact name